File: app/src/main/java/net/opendasharchive/openarchive/util/Prefs.java

package net.opendasharchive.openarchive.util;

import android.content.Context;
import android.content.SharedPreferences;

import androidx.preference.PreferenceManager;

public class Prefs {

    public final static String PREF_UPLOAD_WIFI_ONLY = "upload_wifi_only";
    public final static String PREF_NEARBY_USE_BLUETOOTH = "nearby_use_bluetooth";
    public final static String PREF_NEARBY_USE_WIFI = "nearby_use_wifi";
    public final static String PREF_USE_TOR = "use_tor";
    public final static String PREF_USE_PROOFMODE = "use_proofmode";
    public final static String PREF_USE_NEXTCLOUD_CHUNKING = "upload_nextcloud_chunks";

    public final static String PREF_CURRENT_SPACE_ID = "current_space";

    private static SharedPreferences prefs;

    public static void setContext(Context context) {
        if (prefs == null)
            prefs = PreferenceManager.getDefaultSharedPreferences(context);

    }

    public static void putBoolean(String key, boolean value) {
        prefs.edit().putBoolean(key, value).apply();
    }

    public static boolean getBoolean (String key) {
        return prefs.getBoolean(key,false);
    }


    public static void putString(String key, String value) {
        prefs.edit().putString(key, value).apply();
    }

    public static boolean useNextcloudChunking ()
    {
        return prefs.getBoolean(PREF_USE_NEXTCLOUD_CHUNKING, false);
    }
    public static boolean getUploadWifiOnly ()
    {
    	return prefs.getBoolean(PREF_UPLOAD_WIFI_ONLY, false);
    }
    
    public static void setUploadWifiOnly (boolean wifiOnly)
    {
    	putBoolean(PREF_UPLOAD_WIFI_ONLY,wifiOnly);
    }

    public static boolean getNearbyUseBluetooth ()
    {
        return prefs.getBoolean(PREF_NEARBY_USE_BLUETOOTH, false);
    }

    public static void setNearbyUseBluetooth (boolean useBluetooth)
    {
        putBoolean(PREF_NEARBY_USE_BLUETOOTH,useBluetooth);
    }

    public static boolean getNearbyUseWifi ()
    {
        return prefs.getBoolean(PREF_NEARBY_USE_WIFI, false);
    }

    public static void setNearbyUseWifi (boolean useWifi)
    {
        putBoolean(PREF_NEARBY_USE_WIFI,useWifi);
    }

    public static boolean getUseProofMode () { return prefs.getBoolean(PREF_USE_PROOFMODE, false);}
    public static boolean getUseTor ()
    {
        return prefs.getBoolean(PREF_USE_TOR, false);
    }

    public static void setUseTor (boolean useTor)
    {
        putBoolean(PREF_USE_TOR,useTor);
    }

    public static long getCurrentSpaceId () { return prefs.getLong(PREF_CURRENT_SPACE_ID, -1L);}

    public static void setCurrentSpaceId (long spaceId)
    {
        prefs.edit().putLong(PREF_CURRENT_SPACE_ID, spaceId).commit();
    }

    public static final String PREF_HAS_CONSENT = "ci-consented";

    public static void giveConsent ()
    {
        putBoolean(PREF_HAS_CONSENT,true);
    }
    public static boolean hasConsent ()
    {
        return prefs.getBoolean(PREF_HAS_CONSENT, false);
    }
}

File: app/src/main/java/net/opendasharchive/openarchive/util/Prefs.kt
package net.opendasharchive.openarchive.util

import android.content.Context
import android.content.SharedPreferences
import android.util.Base64
import androidx.preference.PreferenceManager
import org.witness.proofmode.ProofMode

object Prefs {

    private const val UPLOAD_WIFI_ONLY = "upload_wifi_only"
    private const val NEARBY_USE_BLUETOOTH = "nearby_use_bluetooth"
    private const val NEARBY_USE_WIFI = "nearby_use_wifi"
    const val USE_TOR = "use_tor"
    const val USE_PROOFMODE = "use_proofmode"
    private const val USE_NEXTCLOUD_CHUNKING = "upload_nextcloud_chunks"
    const val THEME = "theme"
    private const val CURRENT_SPACE_ID = "current_space"
    private const val FLAG_HINT_SHOWN = "ft.flag"
    private const val BATCH_HINT_SHOWN = "ft.batch"
    private const val IA_HINT_SHOWN = "ft.ia"
    private const val TRACK_LOCATION = "trackLocation"
    private const val NEXTCLOUD_USER_DATA = "next_cloud_user_data"
    private const val LICENSE_URL = "archive_pref_share_license_url"
    private const val PROOFMODE_ENCRYPTED_PASSPHRASE = "proof_mode_encrypted_passphrase"

    private var prefs: SharedPreferences? = null

    fun load(context: Context) {
        if (prefs == null) prefs = PreferenceManager.getDefaultSharedPreferences(context)
    }

    var nextCloudModel: String?
        get() = prefs?.getString(NEXTCLOUD_USER_DATA, "") ?: ""
        set(value) {
            prefs?.edit()?.putString(NEXTCLOUD_USER_DATA, value)?.apply()
        }

    val useNextcloudChunking: Boolean
        get() = prefs?.getBoolean(USE_NEXTCLOUD_CHUNKING, false) ?: false

    var uploadWifiOnly: Boolean
        get() = prefs?.getBoolean(UPLOAD_WIFI_ONLY, false) ?: false
        set(value) {
            prefs?.edit()?.putBoolean(UPLOAD_WIFI_ONLY, value)?.apply()
        }

    var nearbyUseBluetooth: Boolean
        get() = prefs?.getBoolean(NEARBY_USE_BLUETOOTH, false) ?: false
        set(value) {
            prefs?.edit()?.putBoolean(NEARBY_USE_BLUETOOTH, value)?.apply()
        }

    var nearbyUseWifi: Boolean
        get() = prefs?.getBoolean(NEARBY_USE_WIFI, false) ?: false
        set(value) {
            prefs?.edit()?.putBoolean(NEARBY_USE_WIFI, value)?.apply()
        }

    val useProofMode: Boolean
        get() = prefs?.getBoolean(USE_PROOFMODE, false) ?: false

    var useTor: Boolean
        get() = prefs?.getBoolean(USE_TOR, false) ?: false
        set(value) {
            prefs?.edit()?.putBoolean(USE_TOR, value)?.apply()
        }

    var currentSpaceId: Long
        get() = prefs?.getLong(CURRENT_SPACE_ID, -1) ?: -1
        set(value) {
            prefs?.edit()?.putLong(CURRENT_SPACE_ID, value)?.apply()
        }

    var flagHintShown: Boolean
        get() = prefs?.getBoolean(FLAG_HINT_SHOWN, false) ?: false
        set(value) {
            prefs?.edit()?.putBoolean(FLAG_HINT_SHOWN, value)?.apply()
        }

    var batchHintShown: Boolean
        get() = prefs?.getBoolean(BATCH_HINT_SHOWN, false) ?: false
        set(value) {
            prefs?.edit()?.putBoolean(BATCH_HINT_SHOWN, value)?.apply()
        }

    var iaHintShown: Boolean
        get() = prefs?.getBoolean(IA_HINT_SHOWN, false) ?: false
        set(value) {
            prefs?.edit()?.putBoolean(IA_HINT_SHOWN, value)?.apply()
        }

    var trackLocation: Boolean
        get() = prefs?.getBoolean(TRACK_LOCATION, false) ?: false
        set(value) {
            prefs?.edit()?.putBoolean(TRACK_LOCATION, value)?.apply()
        }

    var proofModeLocation: Boolean
        get() = prefs?.getBoolean(ProofMode.PREF_OPTION_LOCATION, false) ?: false
        set(value) {
            prefs?.edit()?.putBoolean(ProofMode.PREF_OPTION_LOCATION, value)?.apply()
        }

    var proofModeNetwork: Boolean
        get() = prefs?.getBoolean(ProofMode.PREF_OPTION_NETWORK, false) ?: false
        set(value) {
            prefs?.edit()?.putBoolean(ProofMode.PREF_OPTION_NETWORK, value)?.apply()
        }

    var licenseUrl: String?
        get() = prefs?.getString(LICENSE_URL, null)
        set(value) {
            prefs?.edit()?.putString(LICENSE_URL, value)?.apply()
        }

    var proofModeEncryptedPassphrase: ByteArray?
        get() {
            val passphrase = prefs?.getString(PROOFMODE_ENCRYPTED_PASSPHRASE, null) ?: return null

            return Base64.decode(passphrase, Base64.DEFAULT)
        }
        set(value) {
            val passphrase =
                if (value == null) null else Base64.encodeToString(value, Base64.DEFAULT)

            prefs?.edit()?.putString(PROOFMODE_ENCRYPTED_PASSPHRASE, passphrase)?.apply()
        }

    val theme: Theme
        get() = Theme.get(prefs?.getString(THEME, null))
}
